src/pyroboplan/planning/prm.py

```diff
@@ -0,0 +1,285 @@
+
+        # Attempt to connect at most `max_neighbor_connections` neighbors.
+        success = False
+        for node, _ in neighbors[0 : self.options.max_neighbor_connections]:
+            # If the nodes are connectable then add an edge.
+            if has_collision_free_path(
+                node.q,
+                new_node.q,
+                self.options.max_angle_step,
+                self.model,
+                self.collision_model,
+            ):
+                self.graph.add_edge(node, new_node)
+                success |= True
+
+        return success
+
+    def reset(self):
+        """
+        Resets the PRM's transient data between queries.
+        """
+        self.latest_path = None
+        for node in self.graph.nodes:
+            node.parent = None
+            node.cost = None
+
+    def plan(self, q_start, q_goal):
+        """
+        Plans a path from a start to a goal configuration using the constructed graph.
+
+        Parameters
+        ----------
+            q_start : array-like
+                The starting robot configuration.
+            q_goal : array-like
+                The goal robot configuration.
+
+        Returns
+        -------
+            list[array-like] :
+                A path from the start to the goal state, if one exists. Otherwise None.
+        """
+
+        # Check start and end pose collisions.
+        if check_collisions_at_state(self.model, self.collision_model, q_start):
+            print("Start configuration in collision.")
+            return None
+        if check_collisions_at_state(self.model, self.collision_model, q_goal):
+            print("Goal configuration in collision.")
+            return None
+
+        try:
+            # Ensure the start and goal nodes are in the graph.
+            start_node = Node(q_start)
+            self.graph.add_node(start_node)
+            goal_node = Node(q_goal)
+            self.graph.add_node(goal_node)
+
+            # If we cannot connect the start and goal nodes then there is no recourse.
+            if not self.connect_node(start_node, self.options.max_neighbor_radius):
+                print("Failed to connect the start configuration to the PRM.")
+                return None
+            if not self.connect_node(goal_node, self.options.max_neighbor_radius):
+                print("Failed to connect the goal configuration to the PRM.")
+                return None
+
+            # Use a graph search to determine if there is a path between the start and goal poses.
+            node_path = astar(self.graph, start_node, goal_node)
+
+            # Reconstruct the path if it exists
+            path = [node.q for node in node_path] if node_path else None
+            self.latest_path = path
+
+        # Always the start and end nodes from the PRM.
+        finally:
+            self.graph.remove_node(start_node)
+            self.graph.remove_node(goal_node)
```

Comment:
Why is this in a try-catch? Can't we just do this at the end of a loop?

Reply:
Ah yeah, just in a `finally` block. To make sure that the nodes are always removed no matter what. I don't think there are exceptions (ha) in the loop, but it will definitely be executed before the return statements above - since they're always added at the start of the function.

Reply:
I would say yes, since it would be unexpected for the map to keep growing in memory during usage.

Reply:
Checks out. In that case I think the finally is the safest bet? Or rearrange the connectivity checks.

Reply:
Agh, I see, so it's not even that it can throw. We're just taking advantage of the "finally".

I would still prefer to rearrange the code without needing to rely on try/finally, but I think I've beaten this PR up pretty bad already, so I won't block.

Reply:
Lol. It's your repo! Happy to fix it.